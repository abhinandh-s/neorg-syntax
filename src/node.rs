#![allow(dead_code)]

use std::sync::Arc;

use crate::{Span, SyntaxKind, Token, token};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SyntaxNode(Repr);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum Repr {
    Leaf(LeafNode),
    Inner(Arc<InnerNode>),
    Error(Arc<ErrorNode>),
}

impl SyntaxNode {
    /// Create a new leaf node.
    pub fn leaf(tok: Token) -> Self {
        Self(Repr::Leaf(LeafNode::new(tok)))
    }

    /// Create a new inner node with children.
    pub fn inner(kind: SyntaxKind, children: Vec<SyntaxNode>) -> Self {
        Self(Repr::Inner(Arc::new(InnerNode::new(kind, children))))
    }

    /// Create a new error node.
    pub fn error(error: SyntaxError, text: impl Into<String>) -> Self {
        Self(Repr::Error(Arc::new(ErrorNode::new(error, text))))
    }

    /// Create a dummy node of the given kind.
    ///
    /// Panics if `kind` is `SyntaxKind::Error`.
    #[track_caller]
    pub const fn placeholder(kind: SyntaxKind) -> Self {
        if matches!(kind, SyntaxKind::Error) {
            panic!("cannot create error placeholder");
        }
        Self(Repr::Leaf(LeafNode {
            kind,
            text: String::new(),
            span: Span::detached(),
        }))
    }

    /// The type of the node.
    pub fn kind(&self) -> SyntaxKind {
        match &self.0 {
            Repr::Leaf(leaf) => leaf.kind,
            Repr::Inner(inner) => inner.kind,
            Repr::Error(_) => SyntaxKind::Error,
        }
    }

    /// The span of the node.
    pub fn span(&self) -> Span {
        match &self.0 {
            Repr::Leaf(leaf) => leaf.span,
            Repr::Inner(inner) => inner.span,
            Repr::Error(node) => node.error.span,
        }
    }

    /// The text of the node if it is a leaf or error node.
    ///
    /// Returns the empty string if this is an inner node.
    pub fn text(&self) -> &String {
        static EMPTY: String = String::new();
        match &self.0 {
            Repr::Leaf(leaf) => &leaf.text,
            Repr::Inner(_) => &EMPTY,
            Repr::Error(node) => &node.text,
        }
    }

    /// Extract the text from the node.
    ///
    /// Builds the string if this is an inner node.
    pub fn into_text(self) -> String {
        match self.0 {
            Repr::Leaf(leaf) => leaf.text,
            Repr::Inner(inner) => inner
                .children
                .iter()
                .cloned()
                .map(Self::into_text)
                .collect(),
            Repr::Error(node) => node.text.clone(),
        }
    }

    /// The node's children.
    pub fn children(&self) -> std::slice::Iter<'_, SyntaxNode> {
        match &self.0 {
            Repr::Leaf(_) | Repr::Error(_) => [].iter(),
            Repr::Inner(inner) => inner.children.iter(),
        }
    }

    /// Whether the node or its children contain an error.
    pub fn erroneous(&self) -> bool {
        match &self.0 {
            Repr::Leaf(_) => false,
            Repr::Inner(inner) => inner.erroneous,
            Repr::Error(_) => true,
        }
    }

    /// The error messages for this node and its descendants.
    pub fn errors(&self) -> Vec<SyntaxError> {
        if !self.erroneous() {
            return vec![];
        }

        if let Repr::Error(node) = &self.0 {
            vec![node.error.clone()]
        } else {
            self.children()
                .filter(|node| node.erroneous())
                .flat_map(|node| node.errors())
                .collect()
        }
    }
    /// Add a user-presentable hint if this is an error node.
    pub fn hint(&mut self, hint: impl Into<String>) {
        if let Repr::Error(node) = &mut self.0 {
            Arc::make_mut(node).hint(hint);
        }
    }

    /// Convert the child to an error, if it isn't already one.
    pub(super) fn convert_to_error(&mut self, message: impl Into<String>) {
        if !self.kind().is_error() {
            let text = std::mem::take(self).into_text();
            *self = SyntaxNode::error(SyntaxError::new(message), text);
        }
    }
    /// Convert the child to an error stating that the given thing was
    /// expected, but the current kind was found.
    pub(super) fn expected(&mut self, expected: &str) {
        let kind = self.kind();
        self.convert_to_error(format!("expected {expected}, found {}", kind.text()));
    }

    /// Convert the child to an error stating it was unexpected.
    pub(super) fn unexpected(&mut self) {
        self.convert_to_error(format!("unexpected {}", self.kind().text()));
    }
}

// for 'convert_to_error'
impl Default for SyntaxNode {
    fn default() -> Self {
        Self::leaf(token!(SyntaxKind::Eof, "", 0))
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
struct LeafNode {
    /// What kind of node this is (each kind would have its own struct in a
    /// strongly typed AST).
    kind: SyntaxKind,
    /// The source text of the node.
    text: String,
    /// The node's span.
    span: Span,
}

impl LeafNode {
    /// Create a new leaf node.
    #[track_caller]
    fn new(token: Token) -> Self {
        debug_assert!(!token.kind().is_error());
        Self {
            kind: token.kind(),
            text: token.text().into(),
            span: Span {
                start: token.len(),
                end: token.len() + token.len(),
            },
        }
    }

    /// The byte length of the node in the source text.
    fn len(&self) -> usize {
        self.text.len()
    }

    /// Whether the two leaf nodes are the same apart from spans.
    fn spanless_eq(&self, other: &Self) -> bool {
        self.kind == other.kind && self.text == other.text
    }
}

impl std::fmt::Debug for LeafNode {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}: {:?}", self.kind, self.text)
    }
}

/// An inner node in the untyped syntax tree.
#[derive(Clone, Eq, PartialEq, Hash)]
struct InnerNode {
    /// What kind of node this is (each kind would have its own struct in a
    /// strongly typed AST).
    kind: SyntaxKind,
    /// The node's span.
    span: Span,
    /// Whether this node or any of its children are erroneous.
    erroneous: bool,
    /// This node's children, losslessly make up this node.
    children: Vec<SyntaxNode>,
}

impl InnerNode {
    /// Create a new inner node with the given kind and children.
    #[track_caller]
    fn new(kind: SyntaxKind, children: Vec<SyntaxNode>) -> Self {
        debug_assert!(!kind.is_error());

        let mut erroneous = false;

        for child in &children {
            erroneous |= child.erroneous();
        }

        Self {
            kind,
            span: Span::default(),
            erroneous,
            children,
        }
    }
}

impl std::fmt::Debug for InnerNode {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self.kind)?;
        if !self.children.is_empty() {
            f.write_str(" ")?;
            f.debug_list().entries(&self.children).finish()?;
        }
        Ok(())
    }
}

/// An error node in the untyped syntax tree.
#[derive(Clone, Eq, PartialEq, Hash)]
struct ErrorNode {
    /// The source text of the node.
    text: String,
    /// The syntax error.
    error: SyntaxError,
}

impl ErrorNode {
    /// Create new error node.
    fn new(error: SyntaxError, text: impl Into<String>) -> Self {
        Self {
            text: text.into(),
            error,
        }
    }

    /// The byte length of the node in the source text.
    fn len(&self) -> usize {
        self.text.len()
    }

    /// Add a user-presentable hint to this error node.
    fn hint(&mut self, hint: impl Into<String>) {
        self.error.hints.push(hint.into());
    }

    /// Whether the two leaf nodes are the same apart from spans.
    fn spanless_eq(&self, other: &Self) -> bool {
        self.text == other.text && self.error.spanless_eq(&other.error)
    }
}

impl std::fmt::Debug for ErrorNode {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "Error: {:?} ({})", self.text, self.error.message)
    }
}

/// A syntactical error.
#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct SyntaxError {
    /// The node's span.
    pub span: Span,
    /// The error message.
    pub message: String,
    /// Additional hints to the user, indicating how this error could be avoided
    /// or worked around.
    pub hints: Vec<String>,
}

impl SyntaxError {
    /// Create a new detached syntax error.
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            span: Span::default(),
            message: message.into(),
            hints: vec![],
        }
    }

    /// Whether the two errors are the same apart from spans.
    fn spanless_eq(&self, other: &Self) -> bool {
        self.message == other.message && self.hints == other.hints
    }
}

impl SyntaxNode {
    pub fn display(&self) -> String {
        fn fmt(node: &SyntaxNode, indent: usize, output: &mut String) {
            let padding = "  ".repeat(indent);
            match &node.0 {
                Repr::Leaf(leaf) => {
                    output.push_str(&format!("{padding}{:?}: {:?}\n", leaf.kind, leaf.text));
                }
                Repr::Inner(inner) => {
                    output.push_str(&format!("{padding}{:?}\n", inner.kind));
                    for child in &inner.children {
                        fmt(child, indent + 1, output);
                    }
                }
                Repr::Error(err) => {
                    output.push_str(&format!(
                        "{padding}Error: {:?} ({})\n",
                        err.text, err.error.message
                    ));
                }
            }
        }

        let mut out = String::new();
        fmt(self, 0, &mut out);
        out
    }
}

