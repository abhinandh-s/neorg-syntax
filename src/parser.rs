#![allow(dead_code)]
#![deny(clippy::print_stdout, clippy::print_stderr)]

use crate::node::SyntaxNode;
use crate::*;

/// A marker representing a node's position in the parser. Mainly used for
/// wrapping, but can also index into the parser to access the node, like
/// `p[m]`.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
struct Marker(usize);

// Index into the parser with markers.
impl std::ops::Index<Marker> for Parser {
    type Output = SyntaxNode;

    fn index(&self, m: Marker) -> &Self::Output {
        &self.nodes[m.0]
    }
}

impl std::ops::IndexMut<Marker> for Parser {
    fn index_mut(&mut self, m: Marker) -> &mut Self::Output {
        &mut self.nodes[m.0]
    }
}

pub struct Parser {
    tokens: Vec<Token>,
    cursor: usize,
    /// Whether the parser has the expected set of open/close delimiters. This
    /// only ever transitions from `true` to `false`.
    balanced: bool,
    /// Nodes representing the concrete syntax tree of previously parsed text.
    /// In Code and Math, includes previously parsed trivia, but not `token`.
    nodes: Vec<SyntaxNode>,
}

impl Parser {
    pub fn new(source: &str) -> Self {
        let tokens = Lexer::new(source).lex();
        Self {
            tokens,
            cursor: 0,
            balanced: true,
            nodes: Vec::new(),
        }
    }

    /// A marker that will point to the current token in the parser once it's
    /// been eaten.
    fn start(&self) -> Marker {
        Marker(self.nodes.len())
    }

    #[track_caller]
    fn assert(&mut self, kind: SyntaxKind) {
        assert_eq!(self.current(), kind);
        self.eat();
    }

    fn current(&self) -> SyntaxKind {
        self.tokens[self.cursor].kind()
    }

    fn next(&self) -> Option<SyntaxKind> {
        self.tokens.get(self.cursor + 1).map(|f| f.kind())
    }

    fn eat(&mut self) {
        let node = SyntaxNode::leaf(self.tokens[self.cursor].clone());
        self.nodes.push(node);
        self.cursor += 1;
    }

    fn eat_many(&mut self, eatable: SyntaxSet) {
        while self.at_set(eatable) {
            self.eat();
        }
    }

    /// Eat the token if at `kind`. Returns `true` if eaten.
    ///
    /// Note: In Math and Code, this will ignore trivia in front of the
    /// `kind`, To forbid skipping trivia, consider using `eat_if_direct`.
    fn eat_if(&mut self, kind: SyntaxKind) -> bool {
        let at = self.at(kind);
        if at {
            self.eat();
        }
        at
    }

    /// Consume the given `kind` or produce an error.
    fn expect(&mut self, kind: SyntaxKind) -> bool {
        let at = self.at(kind);
        if at {
            self.eat();
        } else {
            // FIX: eat?
            //  self.nodes().push(SyntaxNode::error(
            //      SyntaxError::new(format!("expected {}, found {}", kind, self.current().text())),
            //      self.current().text(),
            //  ));
            //  self.eat();
        }
        at
    }

    /// Consume the given closing delimiter or produce an error for the matching
    /// opening delimiter at `open`.
    #[track_caller]
    fn expect_closing_delimiter(&mut self, open: Marker, kind: SyntaxKind) {
        if !self.eat_if(kind) {
            self.nodes[open.0].convert_to_error("unclosed delimiter");
        }
    }

    /// Eat the current node and return a reference for in-place mutation.
    #[track_caller]
    fn eat_and_get(&mut self) -> &mut SyntaxNode {
        let offset = self.nodes.len();
        self.eat();
        &mut self.nodes[offset]
    }
    /// Consume the next token (if any) and produce an error stating that it was
    /// unexpected.
    fn unexpected(&mut self) {
        self.eat_and_get().unexpected();
    }
    /// Produce an error that the given `thing` was expected at the position
    /// of the marker `m`.
    fn expected_at(&mut self, m: Marker, thing: &str) {
        let error = SyntaxNode::error(SyntaxError::new(format!("expected {thing}")), "");
        self.nodes.insert(m.0, error);
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn at_set(&self, set: SyntaxSet) -> bool {
        set.contains(self.current())
    }

    fn wrap(&mut self, m: Marker, kind: SyntaxKind) {
        let drained = self.nodes.drain(m.0..).collect();
        dbg!("called wrap");
        tracing::debug!(?drained, "wrap node");
        let node = SyntaxNode::inner(kind, drained);
        tracing::debug!(?m, ?kind, "wrap node");
        self.nodes.push(node);
    }

    pub fn nodes(&self) -> Vec<SyntaxNode> {
        self.nodes.clone()
    }
}

#[tracing::instrument(skip_all)]
fn parse_variable(p: &mut Parser) {
    let m = p.start();
    p.assert(T!('&'));
    if p.current() == SyntaxKind::WhiteSpace {
        p.unexpected();
    }
    parse_atmod_text_chunk(p);

    p.expect_closing_delimiter(m, T!('&'));
    p.wrap(m, SyntaxKind::Variable);
}

#[tracing::instrument(skip_all)]
fn parse_inline_maths(p: &mut Parser) {
    let m = p.start();
    p.assert(T!('$'));
    if p.current() == SyntaxKind::WhiteSpace {
        p.unexpected();
    }
    parse_atmod_text_chunk(p);

    p.expect_closing_delimiter(m, T!('$'));
    p.wrap(m, SyntaxKind::InlineMath);
}

#[tracing::instrument(skip_all)]
fn parse_inline_code(p: &mut Parser) {
    let m = p.start();
    p.assert(T!('`'));
    if p.current() == SyntaxKind::WhiteSpace {
        p.unexpected();
    }
    parse_atmod_text_chunk(p);

    p.expect_closing_delimiter(m, T!('`'));
    p.wrap(m, SyntaxKind::InlineCode);
}

#[tracing::instrument(skip_all)]
fn parse_subscript(p: &mut Parser) {
    let m = p.start();
    p.assert(T!(','));
    if p.current() == SyntaxKind::WhiteSpace {
        p.unexpected();
    }
    parse_atmod_text_chunk(p);

    p.expect_closing_delimiter(m, T!(','));
    p.wrap(m, SyntaxKind::Subscript);
}

#[tracing::instrument(skip_all)]
fn parse_superscript(p: &mut Parser) {
    let m = p.start();
    p.assert(T!('^'));
    if p.current() == SyntaxKind::WhiteSpace {
        p.unexpected();
    }
    parse_atmod_text_chunk(p);

    p.expect_closing_delimiter(m, T!('^'));
    p.wrap(m, SyntaxKind::Superscript);
}

define_attached_modifier!(parse_bold, Asterisk, Bold);
define_attached_modifier!(parse_underline, Underscore, UnderLine);
define_attached_modifier!(parse_strike_through, Hyphen, StrikeThrough);
define_attached_modifier!(parse_spoiler, Exclamation, Spoiler);
define_attached_modifier!(parse_null_modifier, Percent, NullModifier);
define_attached_modifier!(parse_italics, Slash, Italics);

assert_tree!(
    // [case:1/9] perfect bold
    italics_01,
    parse_italics,
    "/an italics text chunk/ blah blah blah"
);

assert_tree!(
    // [case:2/9] not so perfect italics
    // error: Unclosed delimiter
    italics_02,
    parse_italics,
    "/an italics text chunk blah blah blah"
);

assert_tree!(
    // [case:3/9] not so perfect italics
    // error: WhiteSpace at beginning
    italics_03,
    parse_italics,
    "/ an italics text chunk/ blah blah blah"
);

assert_tree!(
    // [case:4/9] not so perfect italics
    // error: WhiteSpace at end
    italics_04,
    parse_italics,
    "/an italics text chunk / blah blah blah"
);

assert_tree!(
    // [case:5/9] not so perfect italics
    // error: WhiteSpace at both end
    italics_05,
    parse_italics,
    "/ an italics text chunk / blah blah blah"
);

assert_tree!(
    // [case:6/9] not so perfect italics
    // error: `LineEnding` inside text chunk
    italics_06,
    parse_italics,
    "/an italics \n text chunk/ blah blah blah"
);

assert_tree!(
    // [case:7/9] not so perfect italics
    // feat: other inline elements inside this ATACHED_MODIFIERS
    italics_07,
    parse_italics,
    "/an italics text chunk and it have a | verbatim | chunk in it :)/ blah blah blah"
);

assert_tree!(
    // [case:8/9] not so perfect italics
    // feat: /this/is still italics/ - cuz no space after `/`
    italics_08,
    parse_italics,
    "/an italics text chunk and it have a | verbatim | chunk in it :)/blah/ blah blah"
);
assert_tree!(
    // [case:9/9] not so perfect italics
    // feat: /this/ is not fully italics/ - cuz space after `/`
    italics_09,
    parse_italics,
    "/an italics text chunk and it have a | verbatim | chunk in it :)/ blah/ blah blah"
);

#[macro_export]
macro_rules! define_attached_modifier {
    ($fn_name:ident, $kind:ident, $wrapin:ident) => {
        #[tracing::instrument(skip_all)]
        fn $fn_name(p: &mut Parser) {
            let m = p.start();
            p.assert(SyntaxKind::$kind);
            if p.current() == SyntaxKind::WhiteSpace {
                p.unexpected();
            }
            time_bound_while!(!p.at_set(syntax_set!(Eof)), {
                if p.at_set(ATACHED_MODIFIERS.add(SyntaxKind::Pipe)) {
                    if p.at(SyntaxKind::$kind) && (p.next() == Some(SyntaxKind::Word)) {
                        p.eat();
                        continue;
                    } else if p.at(SyntaxKind::$kind) && (p.next() == Some(SyntaxKind::WhiteSpace))
                    {
                        break;
                    }
                    parse_attached_modifiers(p);
                } else {
                    parse_atmod_text_chunk(p);
                }
            });

            assert!(
                p.at(SyntaxKind::$kind) && (p.next() == Some(SyntaxKind::Word))
                    || p.at_set(ATACHED_MODIFIERS.add(SyntaxKind::Eof))
            );
            p.expect_closing_delimiter(m, SyntaxKind::$kind);
            p.wrap(m, SyntaxKind::$wrapin);
        }
    };
}

assert_tree!(
    // [case:1/7] perfect bold
    bold_01,
    parse_bold,
    "*a bold text chunk* blah blah blah"
);

assert_tree!(
    // [case:2/7] not so perfect bold
    // error: Unclosed delimiter
    bold_02,
    parse_bold,
    "*a bold text chunk blah blah blah"
);

assert_tree!(
    // [case:3/7] not so perfect bold
    // error: WhiteSpace at beginning
    bold_03,
    parse_bold,
    "* a bold text chunk* blah blah blah"
);

assert_tree!(
    // [case:4/7] not so perfect bold
    // error: WhiteSpace at end
    bold_04,
    parse_bold,
    "*a bold text chunk * blah blah blah"
);

assert_tree!(
    // [case:5/7] not so perfect bold
    // error: WhiteSpace at both end
    bold_05,
    parse_bold,
    "* a bold text chunk * blah blah blah"
);

assert_tree!(
    // [case:6/7] not so perfect bold
    // error: `LineEnding` inside text chunk
    bold_06,
    parse_bold,
    "*a bold \n text chunk* blah blah blah"
);

assert_tree!(
    // [case:7/7] not so perfect bold
    // feat: other inline elements inside this ATACHED_MODIFIERS
    bold_07,
    parse_bold,
    "*a bold text chunk and it have a | verbatim | chunk in it :)* blah blah blah"
);

/// # Deals with:
///
/// - normal TextChunk
/// - inline TextChunk
/// - inline elements
///
///
/// # Paragraph Break
///
/// A paragraph break is defined as an _empty line_. In the simplest case that means two consecutive
/// `line endings` but since Neorg is a /free-form/ markup language, a line which only contains
/// whitespace is also considered empty.
#[tracing::instrument(skip_all)]
pub fn parse_paragraph(p: &mut Parser) {
    let m = p.start();

    time_bound_while!(!p.at(SyntaxKind::Eof), {
        // a paragraph segment stops at `LineEnding`.
        // ie, after a loop if p is at `LineEnding` or `WhiteSpace`
        if let Some(k) = p.next() {
            if p.at(SyntaxKind::LineEnding) {
                let m = p.start();
                p.eat();
                p.wrap(m, SyntaxKind::ParaBreak);
            } else if p.at(SyntaxKind::WhiteSpace) && k == SyntaxKind::LineEnding {
                let m = p.start();
                p.eat();
                p.eat();
                p.wrap(m, SyntaxKind::ParaBreak);
            } else {
                // still p is at WhiteSpace
                parse_para_segment(p);
            }
        }
    });
    p.wrap(m, SyntaxKind::Paragraph);
}

/// # Verbatim Paragraph Segments
///  
/// These are structurally equivalent to regular `paragraph segments` with a single exception.
/// Verbatim paragraph segments are built up from _only_ `words`. This means that attached
/// modifiers and linkables are simply parsed as raw text within a verbatim paragraph segment.
#[tracing::instrument(skip_all)]
fn parse_verbatim(p: &mut Parser) {
    let m = p.start();
    p.assert(T!('|'));

    parse_verbatim_chunk(p);

    p.expect_closing_delimiter(m, T!('|'));
    p.wrap(m, SyntaxKind::Verbatim);
}

#[tracing::instrument(skip_all)]
fn parse_verbatim_chunk(p: &mut Parser) {
    let m = p.start();
    let set = syntax_set!(Pipe, Eof);
    while !p.at_set(set) {
        p.eat();
    }
    p.wrap(m, SyntaxKind::TextChunk);
}

#[tracing::instrument(skip_all)]
fn parse_nm_text_chunk(p: &mut Parser) {
    let m = p.start();

    while p.at_set(syntax_set!(Word, WhiteSpace, Dot)) {
        p.eat();
    }
    p.wrap(m, SyntaxKind::TextChunk);
}

/// If an Operation in this while loop took more than 1 second it will panic
#[macro_export]
macro_rules! time_bound_while {
    ($cond:expr, $body:block) => {{
        #[cfg(debug_assertions)]
        {
            let start = std::time::Instant::now();
            while $cond {
                if start.elapsed() >= std::time::Duration::from_secs(1) {
                    panic!("Operation took more than 1 sec, possible infinite loop.");
                }
                $body
            }
        }
        #[cfg(not(debug_assertions))]
        {
            while $cond {
                $body
            }
        }
    }};
}

#[tracing::instrument(skip_all)]
fn parse_atmod_text(p: &mut Parser, stop_at: SyntaxSet) {
    let m = p.start();

    while !p.at_set(stop_at.add(SyntaxKind::Eof)) {
        if p.at(SyntaxKind::LineEnding) {
            p.unexpected();
        }
        p.eat();
    }

    if p.at_set(ATACHED_MODIFIERS) {
        if let Some(n) = p.nodes.last_mut() {
            if n.kind() == SyntaxKind::WhiteSpace {
                n.unexpected();
            }
        }
    }

    p.wrap(m, SyntaxKind::TextChunk);
}

#[tracing::instrument(skip_all)]
fn parse_atmod_text_chunk(p: &mut Parser) {
    let m = p.start();
    let set = ATACHED_MODIFIERS.add(SyntaxKind::Pipe);

    while !p.at_set(set.add(SyntaxKind::Eof)) {
        if p.at(SyntaxKind::LineEnding) {
            //           p.balanced = false;
            p.unexpected();
        }
        p.eat();
    }

    if p.at_set(ATACHED_MODIFIERS) {
        //  if p.balanced {
        if let Some(n) = p.nodes.last_mut() {
            if n.kind() == SyntaxKind::WhiteSpace {
                n.unexpected();
            }
        }
    }
    // } else {
    //     p.balanced = true
    // }
    p.wrap(m, SyntaxKind::TextChunk);
}
/// # Paragraph Segments
///
/// `Words` are first combined into *paragraph segments*. A paragraph segment
/// may then contain any inline element of type:
///
/// - attached modifiers
/// - linkables
///
/// Usually, a [`line ending`] terminates the paragraph segment.
/// This means that a paragraph segment is simply a line of text:
///
/// # example
///
/// I am a paragraph segment.
/// I am another paragraph segment.
/// Together we form a paragraph.
///
#[tracing::instrument(skip_all)]
fn parse_para_segment(p: &mut Parser) {
    let m = p.start();

    time_bound_while!(!p.at_set(syntax_set!(LineEnding, Eof)), {
        if p.at_set(ATACHED_MODIFIERS.add(SyntaxKind::Pipe)) {
            parse_attached_modifiers(p);
        } else {
            let amod_stoper = ATACHED_MODIFIERS
                .add(SyntaxKind::Pipe)
                .add(SyntaxKind::LineEnding)
                .add(SyntaxKind::Eof);
            while !p.at_set(amod_stoper) {
                p.eat();
            }
        }
    });
    p.expect(SyntaxKind::LineEnding);
    p.wrap(m, SyntaxKind::ParaSegment);
}

assert_tree!(
    long_para_segment_01,
    parse_para_segment,
    "this is a *bold* /italic/ _underline_   -strike-through- !spoiler! ^superscript^ ,subscript, `inline code` %null modifier% &variable& "
);

/// Their name should be rather self-explanatory - both the opening and closing modifier
/// are _attached_ to one another.
///
/// The following attached modifiers exist and have respective meaning:
/// - *bold*
/// - /italic/
/// - _underline_
/// - -strike-through-
/// - !spoiler!
/// - ^superscript^ (cannot be nested into `subscript`)
/// - ,subscript, (cannot be nested into `superscript`)
/// - `inline code` (disables any nested markup - verbatim)
/// - `%null modifier%`
/// - $f(x) = y$ (verbatim)
/// - &variable& (verbatim)
fn parse_attached_modifiers(p: &mut Parser) {
    dbg!("from parse_attached_modifiers:", p.current());
    match p.current() {
        SyntaxKind::Asterisk => parse_bold(p),
        SyntaxKind::Slash => parse_italics(p),
        SyntaxKind::Underscore => parse_underline(p),
        SyntaxKind::Exclamation => parse_spoiler(p),
        SyntaxKind::Caret => parse_superscript(p),
        SyntaxKind::Comma => parse_subscript(p),
        SyntaxKind::Backtick => parse_inline_code(p),
        SyntaxKind::Percent => parse_null_modifier(p),
        SyntaxKind::Dollar => parse_inline_maths(p),
        SyntaxKind::Ampersand => parse_variable(p),
        SyntaxKind::Pipe => parse_verbatim(p),
        SyntaxKind::Hyphen => parse_strike_through(p),
        _ => panic!("unimplemented inline!"),
    }
}

#[macro_export]
macro_rules! assert_tree {
    ($test_name:ident, $parse_fn:ident, $input:literal) => {
        #[test]
        fn $test_name() {
            let snapshot_path = {
                let root = env!("CARGO_MANIFEST_DIR");
                std::path::Path::new(root)
                    .join("tests")
                    .join("snapshots")
            };

            let mut p = $crate::Parser::new($input);
            $crate::parser::$parse_fn(&mut p);
            assert_eq!(p.nodes().len(), 1);

            let output = p.nodes()[0].display();

            // puts it in tests/snapshots/
            // and do not prepend path before snaps name
            insta::with_settings!({ snapshot_path => snapshot_path, prepend_module_to_snapshot => false }, {
                insta::assert_snapshot!(output);
            });
        }
    };
}

#[cfg(test)]
mod test {
    assert_tree!(verbatim_05, parse_verbatim, "|this is a test");
    assert_tree!(verbatim_01, parse_verbatim, "|this is a test|");
    assert_tree!(verbatim_02, parse_verbatim, "| this is a test |");
    assert_tree!(
        verbatim_03,
        parse_verbatim,
        r###"|~ this is a test !"#$%&'()*+,-./:;<=>?@[]^_`{}~\thre|"###
    );
    assert_tree!(verbatim_04, parse_verbatim, r###"|~ `{}~\thre|"###);
    assert_tree!(text_chunk_01, parse_atmod_text_chunk, "this is text chunk");
    assert_tree!(
        text_chunk_01_err,
        parse_atmod_text_chunk,
        "this is text chunk / not this"
    );
    assert_tree!(
        para_segment_01,
        parse_para_segment,
        "this is |a verbatim text chunk|"
    );
    assert_tree!(
        para_segment_01_err,
        parse_para_segment,
        "this is |a verbatim text chunk"
    );
    assert_tree!(
        para_segment_02_err,
        parse_para_segment,
        "this is |a verbatim text chunk|\nthis is next"
    );
    assert_tree!(
        para_segment_03_err,
        parse_para_segment,
        "this isrbatim text chunk\nthis is next"
    );
    assert_tree!(
        paragraph_01,
        parse_paragraph,
        "this isrbatim text chunk\nthis is next"
    );
    assert_tree!(
        paragraph_02,
        parse_paragraph,
        r#"I am a paragraph segment
I am another paragraph segment
 Together we form a paragraph"#
    );
    assert_tree!(
        paragraph_03,
        parse_paragraph,
        r#"I am a paragraph segment.
I am another paragraph segment.
    Together we form a paragraph."#
    );
    assert_tree!(
        paragraph_with_verbatim_01,
        parse_paragraph,
        r#"I am a paragraph segment.
I am another paragraph segment.
this |is verbatim| content
    Together we form a paragraph."#
    );
    assert_tree!(
        paragraph_with_verbatim_emph_01,
        parse_paragraph,
        r#"I am a paragraph segment.
I am another paragraph segment.
this |is verbatim| content *this is bold*
    Together we form a paragraph."#
    );

    // ParaBreak
    //
    // [case:1/2]
    //
    // `LineEnding` after a `LineEnding`
    assert_tree!(
        parabreak_01,
        parse_paragraph,
        r#" I am a paragraph segment.
I am another paragraph segment.

this |is verbatim| content *this is bold*
    Together we form a paragraph."#
    );

    // ParaBreak
    //
    // [case:2/2]
    //
    // empty line with many `WhiteSpace` followed by `LineEnding`
    assert_tree!(
        parabreak_02,
        parse_paragraph,
        r#" I am a paragraph segment.
I am another paragraph segment.
      
this |is verbatim| content *this is bold*
    Together we form a paragraph."#
    );
    assert_tree!(
        parse_02,
        parse_paragraph,
        r#" I am a paragraph segment.

I am another paragraph segment.
 are _attached_ to one another.
([{}])
  *bold*
  /italic/
  _underline_
  |-strike-through-|
  !spoiler!
  ^superscript^
  ,subscript,
  `inline code`
  `%null modifier%`
  &variable& 
this |is verbatim| content *this is bold*
    Together we form a paragraph."#
    );

    quickcheck::quickcheck! {
        #[test]
        #[ignore]
        fn no_panic_on_random_input(input: String) -> bool {
            let mut parser = crate::Parser::new(&input);
            crate::parse_paragraph(&mut parser);
            true // test passes if no panic
        }
    }

    proptest::proptest! {
        #[test]
        #[ignore]
        fn no_panic_prop(input in ".*") {
            let mut parser = crate::Parser::new(&input);
            crate::parse_paragraph(&mut parser);
        }
    }
}
