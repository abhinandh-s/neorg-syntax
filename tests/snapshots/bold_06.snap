---
source: src/parser.rs
expression: output
---
Bold
  Asterisk: "*"
  TextChunk
    Word: "a"
    WhiteSpace: " "
    Word: "bold"
    WhiteSpace: " "
    Error: "\n" (unexpected LineEnding)
    WhiteSpace: " "
    Word: "text"
    WhiteSpace: " "
    Word: "chunk"
  Asterisk: "*"
